Code action: add additional constraint

I just got:

```
C:\Neil\model\Main.hs:20:19: error:
    • No instance for (Hashable a) arising from a use of ‘hash’
      Possible fix:
        add (Hashable a) to the context of
          the type signature for:
            apply :: forall a. Int -> RW a -> Map.Map a Int -> Map.Map a Int
    • In the expression:
        hash (name, [(v, mp Map.! v) | v <- Set.toAscList $ Set.union r w])
      In an equation for ‘input’:
          input
            = hash
 ```

That could probably have been a code action.

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -472,8 +472,8 @@ suggestInstanceConstraint contents Diagnostic {..} missingConstraint
   | otherwise = []
     where
       findExistingConstraints :: T.Text -> T.Text
-      findExistingConstraints =
-        T.lines >>> flip (!!) 1 >>> T.strip >>> T.replace "from the context: " ""
+      findExistingConstraints t =
+        T.replace "from the context: " "" . T.strip $ T.lines t !! 1
 
       readPositionNumber :: T.Text -> Int
       readPositionNumber = T.unpack >>> read >>> pred
@@ -529,7 +529,7 @@ suggestFunctionConstraint contents Diagnostic{..} missingConstraint
       findExistingConstraints :: T.Text -> Maybe T.Text
       findExistingConstraints message =
         if message =~ ("from the context:" :: String)
-           then matchRegex message "\\. ([^=]+)" <&> head <&> T.strip
+           then fmap (T.strip . head) $ matchRegex message "\\. ([^=]+)"
            else Nothing
 
       buildNewConstraints :: T.Text -> Maybe T.Text -> T.Text

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -56,6 +56,7 @@ import Outputable (ppr, showSDocUnsafe)
 import DynFlags (xFlags, FlagSpec(..))
 import GHC.LanguageExtensions.Type (Extension)
 import System.Time.Extra (showDuration, duration)
+import Data.Function
 
 plugin :: Plugin c
 plugin = codeActionPluginWithRules rules codeAction <> Plugin mempty setHandlersCodeLens
@@ -155,6 +156,7 @@ suggestAction dflags packageExports ideOptions parsedModule text diag = concat
     , suggestModuleTypo diag
     , suggestReplaceIdentifier text diag
     , suggestSignature True diag
+    , suggestInstanceConstraint text diag
     ] ++ concat
     [  suggestNewDefinition ideOptions pm text diag
     ++ suggestRemoveRedundantImport pm text diag
@@ -403,6 +405,26 @@ suggestSignature isQuickFix Diagnostic{_range=_range@Range{..},..}
             = 0
 
 suggestSignature _ _ = []
+
+suggestInstanceConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
+suggestInstanceConstraint contents Diagnostic {..}
+-- • No instance for (Eq a) arising from a use of ‘==’
+--   Possible fix: add (Eq a) to the context of the instance declaration
+-- • In the expression: x == y
+--   In an equation for ‘==’: (Wrap x) == (Wrap y) = x == y
+--   In the instance declaration for ‘Eq (Wrap a)’
+  | Just c <- contents
+  , Just [constraint] <- matchRegex _message "No instance for ([^’]*) arising from a use of"
+  , Just [instanceDeclaration] <- matchRegex _message "In the instance declaration for ‘([^`]*)’"
+  = let instanceLine = c
+            & T.splitOn ("instance " <> instanceDeclaration)
+            & head & T.lines & length
+        startOfConstraint = Position instanceLine (length ("instance " :: String))
+        range = Range startOfConstraint startOfConstraint
+        title = "Add `" <> constraint <> "` to the context of the instance declaration"
+        newConstraint = constraint <> " => "
+     in [(title, [TextEdit range newConstraint])]
+  |  otherwise = []
 
 -------------------------------------------------------------------------------------------------
 

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -421,8 +421,10 @@ normalizeConstraints existingConstraints constraint =
                            else "(" <> existingConstraints
    in constraintsInit <> ", " <> constraint <> ")"
 
+-- | Suggests a constraint for an instance declaration for which a constraint is missing.
 suggestInstanceConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestInstanceConstraint contents Diagnostic {..}
+-- Suggests a constraint for an instance declaration with no existing constraints.
 -- • No instance for (Eq a) arising from a use of ‘==’
 --   Possible fix: add (Eq a) to the context of the instance declaration
 -- • In the expression: x == y
@@ -439,6 +441,8 @@ suggestInstanceConstraint contents Diagnostic {..}
         title = "Add `" <> constraint <> "` to the context of the instance declaration"
         newConstraint = constraint <> " => "
      in [(title, [TextEdit range newConstraint])]
+
+-- Suggests a constraint for an instance declaration with one or more existing constraints.
 -- • Could not deduce (Eq b) arising from a use of ‘==’
 --   from the context: Eq a
 --     bound by the instance declaration at /path/to/Main.hs:7:10-32
@@ -475,8 +479,10 @@ findTypeSignatureLine :: T.Text -> T.Text -> Int
 findTypeSignatureLine contents typeSignatureName =
   T.splitOn (typeSignatureName <> " :: ") contents & head & T.lines & length
 
+-- | Suggests a constraint for a type signature for which a constraint is missing.
 suggestFunctionConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestFunctionConstraint contents Diagnostic{..}
+-- Suggests a constraint for a type signature with any number of existing constraints.
 -- • No instance for (Eq a) arising from a use of ‘==’
 --   Possible fix:
 --     add (Eq a) to the context of

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -473,7 +473,7 @@ suggestInstanceConstraint contents Diagnostic {..}
       readPositionNumber = T.unpack >>> read >>> pred
 
 findTypeSignatureName :: T.Text -> Maybe T.Text
-findTypeSignatureName t = matchRegex t "([a-zA-Z0-9]+) :: " <&> head
+findTypeSignatureName t = matchRegex t "\\n\\s*(.+) :: " <&> head
 
 findTypeSignatureLine :: T.Text -> T.Text -> Int
 findTypeSignatureLine contents typeSignatureName =

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -460,7 +460,7 @@ suggestInstanceConstraint contents Diagnostic {..}
         constraintFirstChar = readPositionNumber constraintFirstCharStr
         startOfConstraint = Position instanceLine constraintFirstChar
         endOfConstraint = Position instanceLine $
-          constraintFirstChar + (T.length existingConstraints)
+          constraintFirstChar + T.length existingConstraints
         range = Range startOfConstraint endOfConstraint
         title = "Add `" <> constraint <> "` to the context of the instance declaration"
      in [(title, [TextEdit range newConstraints])]

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -410,7 +410,16 @@ suggestSignature isQuickFix Diagnostic{_range=_range@Range{..},..}
 suggestSignature _ _ = []
 
 findMissingConstraint :: T.Text -> Maybe T.Text
-findMissingConstraint t = matchRegex t "No instance for \\((.+)\\) arising from a use of" <&> head
+findMissingConstraint t =
+    let regex = "(No instance for|Could not deduce) \\((.+)\\) arising from a use of"
+     in matchRegex t regex <&> last
+
+normalizeConstraints :: T.Text -> T.Text -> T.Text
+normalizeConstraints existingConstraints constraint =
+  let constraintsInit = if "(" `T.isPrefixOf` existingConstraints
+                           then T.dropEnd 1 existingConstraints
+                           else "(" <> existingConstraints
+   in constraintsInit <> ", " <> constraint <> ")"
 
 suggestInstanceConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestInstanceConstraint contents Diagnostic {..}
@@ -438,15 +447,16 @@ suggestInstanceConstraint contents Diagnostic {..}
 --   In the expression: x == y && x' == y'
 --   In an equation for ‘==’:
 --       (Pair x x') == (Pair y y') = x == y && x' == y'
-  | Just [constraint] <- matchRegex _message "Could not deduce \\((.+)\\) arising from a use of"
+  | Just constraint <- findMissingConstraint _message
   , Just [instanceLineStr, constraintFirstCharStr]
     <- matchRegex _message "bound by the instance declaration at .+:([0-9]+):([0-9]+)"
-  = let existingConstraint = secondLine _message
-        newConstraints = normalizeConstraints existingConstraint constraint
+  = let existingConstraints = secondLine _message
+        newConstraints = normalizeConstraints existingConstraints constraint
         instanceLine = readPositionNumber instanceLineStr
         constraintFirstChar = readPositionNumber constraintFirstCharStr
         startOfConstraint = Position instanceLine constraintFirstChar
-        endOfConstraint = Position instanceLine $ constraintFirstChar + (T.length existingConstraint)
+        endOfConstraint = Position instanceLine $
+          constraintFirstChar + (T.length existingConstraints)
         range = Range startOfConstraint endOfConstraint
         title = "Add `" <> constraint <> "` to the context of the instance declaration"
      in [(title, [TextEdit range newConstraints])]
@@ -458,12 +468,12 @@ suggestInstanceConstraint contents Diagnostic {..}
       readPositionNumber :: T.Text -> Int
       readPositionNumber = T.unpack >>> read >>> pred
 
-      normalizeConstraints :: T.Text -> T.Text -> T.Text
-      normalizeConstraints existingConstraint constraint =
-        (if "(" `T.isPrefixOf` existingConstraint
-           then T.dropEnd 1 existingConstraint
-           else "(" <> existingConstraint
-        ) <> ", " <> constraint <> ")"
+findTypeSignatureName :: T.Text -> Maybe T.Text
+findTypeSignatureName t = matchRegex t "([a-zA-Z0-9]+) :: " <&> head
+
+findTypeSignatureLine :: T.Text -> T.Text -> Int
+findTypeSignatureLine contents typeSignatureName =
+  T.splitOn (typeSignatureName <> " :: ") contents & head & T.lines & length
 
 suggestFunctionConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestFunctionConstraint contents Diagnostic{..}
@@ -474,19 +484,50 @@ suggestFunctionConstraint contents Diagnostic{..}
 --         eq :: forall a. a -> a -> Bool
 -- • In the expression: x == y
 --   In an equation for ‘eq’: eq x y = x == y
-  | Just c <- contents
-  , Just [] <- matchRegex _message "In an equation for"
-  , Just constraint <- findMissingConstraint _message
-  , Just [typeSignatureName] <- matchRegex _message "([a-zA-Z0-9]+) :: "
-  = let newConstraint = constraint <> " => "
-        typeSignatureLine = T.splitOn c typeSignatureName & head & T.lines & length & (+1)
-        typeSignatureFirstChar = T.length $ typeSignatureName <> " :: "
-        title = "Add `" <> constraint <> "` to the context of the type signature for `" <> typeSignatureName <> "`"
-        position = Position typeSignatureLine typeSignatureFirstChar
-        range = Range position position
-     in [(title, [TextEdit range newConstraint])]
-  | otherwise = []
 
+-- • Could not deduce (Eq b) arising from a use of ‘==’
+--   from the context: Eq a
+--     bound by the type signature for:
+--                eq :: forall a b. Eq a => Pair a b -> Pair a b -> Bool
+--     at Main.hs:5:1-42
+--   Possible fix:
+--     add (Eq b) to the context of
+--       the type signature for:
+--         eq :: forall a b. Eq a => Pair a b -> Pair a b -> Bool
+-- • In the second argument of ‘(&&)’, namely ‘y == y'’
+--   In the expression: x == x' && y == y'
+--   In an equation for ‘eq’:
+--       eq (Pair x y) (Pair x' y') = x == x' && y == y'
+  | Just c <- contents
+  , True <- _message =~ ("the type signature for:" :: String)
+  , Just constraint <- findMissingConstraint _message
+  , Just typeSignatureName <- findTypeSignatureName _message
+  = let mExistingConstraints = findExistingConstraints _message
+        newConstraint = buildNewConstraints constraint mExistingConstraints
+        typeSignatureLine = findTypeSignatureLine c typeSignatureName
+        typeSignatureFirstChar = T.length $ typeSignatureName <> " :: "
+        startOfConstraint = Position typeSignatureLine typeSignatureFirstChar
+        endOfConstraint = Position typeSignatureLine $
+          typeSignatureFirstChar + maybe 0 T.length mExistingConstraints
+        range = Range startOfConstraint endOfConstraint
+     in [(actionTitle constraint typeSignatureName, [TextEdit range newConstraint])]
+  | otherwise = []
+    where
+      findExistingConstraints :: T.Text -> Maybe T.Text
+      findExistingConstraints message =
+        if message =~ ("from the context:" :: String)
+           then matchRegex message "\\. ([^=]+)" <&> head <&> T.strip
+           else Nothing
+
+      buildNewConstraints :: T.Text -> Maybe T.Text -> T.Text
+      buildNewConstraints constraint mExistingConstraints =
+        case mExistingConstraints of
+          Just existingConstraints -> normalizeConstraints existingConstraints constraint
+          Nothing -> constraint <> " => "
+
+      actionTitle :: T.Text -> T.Text -> T.Text
+      actionTitle constraint typeSignatureName = "Add `" <> constraint
+        <> "` to the context of the type signature for `" <> typeSignatureName <> "`"
 
 -------------------------------------------------------------------------------------------------
 

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -57,6 +57,7 @@ import DynFlags (xFlags, FlagSpec(..))
 import GHC.LanguageExtensions.Type (Extension)
 import System.Time.Extra (showDuration, duration)
 import Data.Function
+import Control.Arrow ((>>>))
 
 plugin :: Plugin c
 plugin = codeActionPluginWithRules rules codeAction <> Plugin mempty setHandlersCodeLens
@@ -414,17 +415,51 @@ suggestInstanceConstraint contents Diagnostic {..}
 --   In an equation for ‘==’: (Wrap x) == (Wrap y) = x == y
 --   In the instance declaration for ‘Eq (Wrap a)’
   | Just c <- contents
-  , Just [constraint] <- matchRegex _message "No instance for ([^’]*) arising from a use of"
+  , Just [constraint] <- matchRegex _message "No instance for \\((.+)\\) arising from a use of"
   , Just [instanceDeclaration] <- matchRegex _message "In the instance declaration for ‘([^`]*)’"
   = let instanceLine = c
-            & T.splitOn ("instance " <> instanceDeclaration)
-            & head & T.lines & length
+          & T.splitOn ("instance " <> instanceDeclaration)
+          & head & T.lines & length
         startOfConstraint = Position instanceLine (length ("instance " :: String))
         range = Range startOfConstraint startOfConstraint
         title = "Add `" <> constraint <> "` to the context of the instance declaration"
         newConstraint = constraint <> " => "
      in [(title, [TextEdit range newConstraint])]
-  |  otherwise = []
+-- • Could not deduce (Eq b) arising from a use of ‘==’
+--   from the context: Eq a
+--     bound by the instance declaration at /path/to/Main.hs:7:10-32
+--   Possible fix: add (Eq b) to the context of the instance declaration
+-- • In the second argument of ‘(&&)’, namely ‘x' == y'’
+--   In the expression: x == y && x' == y'
+--   In an equation for ‘==’:
+--       (Pair x x') == (Pair y y') = x == y && x' == y'
+  | Just [constraint] <- matchRegex _message "Could not deduce \\((.+)\\) arising from a use of"
+  , Just [instanceLineStr, constraintFirstCharStr]
+    <- matchRegex _message "bound by the instance declaration at .+:([0-9]+):([0-9]+)"
+  = let existingConstraint = secondLine _message
+        newConstraints = normalizeConstraints existingConstraint constraint
+        instanceLine = readPositionNumber instanceLineStr
+        constraintFirstChar = readPositionNumber constraintFirstCharStr
+        startOfConstraint = Position instanceLine constraintFirstChar
+        endOfConstraint = Position instanceLine $ constraintFirstChar + (T.length existingConstraint)
+        range = Range startOfConstraint endOfConstraint
+        title = "Add `" <> constraint <> "` to the context of the instance declaration"
+     in [(title, [TextEdit range newConstraints])]
+  | otherwise = []
+    where
+      secondLine :: T.Text -> T.Text
+      secondLine = T.lines >>> flip (!!) 1 >>> T.strip >>> T.replace "from the context: " ""
+
+      readPositionNumber :: T.Text -> Int
+      readPositionNumber = T.unpack >>> read >>> pred
+
+      normalizeConstraints :: T.Text -> T.Text -> T.Text
+      normalizeConstraints existingConstraint constraint =
+        (if "(" `T.isPrefixOf` existingConstraint
+           then T.dropEnd 1 existingConstraint
+           else "(" <> existingConstraint
+        ) <> ", " <> constraint <> ")"
+
 
 -------------------------------------------------------------------------------------------------
 

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -58,6 +58,7 @@ import GHC.LanguageExtensions.Type (Extension)
 import System.Time.Extra (showDuration, duration)
 import Data.Function
 import Control.Arrow ((>>>))
+import Data.Functor
 
 plugin :: Plugin c
 plugin = codeActionPluginWithRules rules codeAction <> Plugin mempty setHandlersCodeLens
@@ -158,6 +159,7 @@ suggestAction dflags packageExports ideOptions parsedModule text diag = concat
     , suggestReplaceIdentifier text diag
     , suggestSignature True diag
     , suggestInstanceConstraint text diag
+    , suggestFunctionConstraint text diag
     ] ++ concat
     [  suggestNewDefinition ideOptions pm text diag
     ++ suggestRemoveRedundantImport pm text diag
@@ -407,6 +409,9 @@ suggestSignature isQuickFix Diagnostic{_range=_range@Range{..},..}
 
 suggestSignature _ _ = []
 
+findMissingConstraint :: T.Text -> Maybe T.Text
+findMissingConstraint t = matchRegex t "No instance for \\((.+)\\) arising from a use of" <&> head
+
 suggestInstanceConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestInstanceConstraint contents Diagnostic {..}
 -- • No instance for (Eq a) arising from a use of ‘==’
@@ -415,7 +420,7 @@ suggestInstanceConstraint contents Diagnostic {..}
 --   In an equation for ‘==’: (Wrap x) == (Wrap y) = x == y
 --   In the instance declaration for ‘Eq (Wrap a)’
   | Just c <- contents
-  , Just [constraint] <- matchRegex _message "No instance for \\((.+)\\) arising from a use of"
+  , Just constraint <- findMissingConstraint _message
   , Just [instanceDeclaration] <- matchRegex _message "In the instance declaration for ‘([^`]*)’"
   = let instanceLine = c
           & T.splitOn ("instance " <> instanceDeclaration)
@@ -459,6 +464,28 @@ suggestInstanceConstraint contents Diagnostic {..}
            then T.dropEnd 1 existingConstraint
            else "(" <> existingConstraint
         ) <> ", " <> constraint <> ")"
+
+suggestFunctionConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
+suggestFunctionConstraint contents Diagnostic{..}
+-- • No instance for (Eq a) arising from a use of ‘==’
+--   Possible fix:
+--     add (Eq a) to the context of
+--       the type signature for:
+--         eq :: forall a. a -> a -> Bool
+-- • In the expression: x == y
+--   In an equation for ‘eq’: eq x y = x == y
+  | Just c <- contents
+  , Just [] <- matchRegex _message "In an equation for"
+  , Just constraint <- findMissingConstraint _message
+  , Just [typeSignatureName] <- matchRegex _message "([a-zA-Z0-9]+) :: "
+  = let newConstraint = constraint <> " => "
+        typeSignatureLine = T.splitOn c typeSignatureName & head & T.lines & length & (+1)
+        typeSignatureFirstChar = T.length $ typeSignatureName <> " :: "
+        title = "Add `" <> constraint <> "` to the context of the type signature for `" <> typeSignatureName <> "`"
+        position = Position typeSignatureLine typeSignatureFirstChar
+        range = Range position position
+     in [(title, [TextEdit range newConstraint])]
+  | otherwise = []
 
 
 -------------------------------------------------------------------------------------------------

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -158,8 +158,7 @@ suggestAction dflags packageExports ideOptions parsedModule text diag = concat
     , suggestModuleTypo diag
     , suggestReplaceIdentifier text diag
     , suggestSignature True diag
-    , suggestInstanceConstraint text diag
-    , suggestFunctionConstraint text diag
+    , suggestConstraint text diag
     ] ++ concat
     [  suggestNewDefinition ideOptions pm text diag
     ++ suggestRemoveRedundantImport pm text diag
@@ -409,10 +408,21 @@ suggestSignature isQuickFix Diagnostic{_range=_range@Range{..},..}
 
 suggestSignature _ _ = []
 
-findMissingConstraint :: T.Text -> Maybe T.Text
-findMissingConstraint t =
-    let regex = "(No instance for|Could not deduce) \\((.+)\\) arising from a use of"
-     in matchRegex t regex <&> last
+-- | Suggests a constraint for a declaration for which a constraint is missing.
+suggestConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
+suggestConstraint mContents diag@Diagnostic {..}
+  | Just contents <- mContents
+  , Just missingConstraint <- findMissingConstraint _message
+  = let codeAction = if _message =~ ("the type signature for:" :: String)
+                        then suggestFunctionConstraint
+                        else suggestInstanceConstraint
+     in codeAction contents diag missingConstraint
+  | otherwise = []
+    where
+      findMissingConstraint :: T.Text -> Maybe T.Text
+      findMissingConstraint t =
+        let regex = "(No instance for|Could not deduce) \\((.+)\\) arising from a use of"
+         in matchRegex t regex <&> last
 
 normalizeConstraints :: T.Text -> T.Text -> T.Text
 normalizeConstraints existingConstraints constraint =
@@ -422,25 +432,22 @@ normalizeConstraints existingConstraints constraint =
    in constraintsInit <> ", " <> constraint <> ")"
 
 -- | Suggests a constraint for an instance declaration for which a constraint is missing.
-suggestInstanceConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
-suggestInstanceConstraint contents Diagnostic {..}
+suggestInstanceConstraint :: T.Text -> Diagnostic -> T.Text -> [(T.Text, [TextEdit])]
+suggestInstanceConstraint contents Diagnostic {..} missingConstraint
 -- Suggests a constraint for an instance declaration with no existing constraints.
 -- • No instance for (Eq a) arising from a use of ‘==’
 --   Possible fix: add (Eq a) to the context of the instance declaration
 -- • In the expression: x == y
 --   In an equation for ‘==’: (Wrap x) == (Wrap y) = x == y
 --   In the instance declaration for ‘Eq (Wrap a)’
-  | Just c <- contents
-  , Just constraint <- findMissingConstraint _message
-  , Just [instanceDeclaration] <- matchRegex _message "In the instance declaration for ‘([^`]*)’"
-  = let instanceLine = c
+  | Just [instanceDeclaration] <- matchRegex _message "In the instance declaration for ‘([^`]*)’"
+  = let instanceLine = contents
           & T.splitOn ("instance " <> instanceDeclaration)
           & head & T.lines & length
         startOfConstraint = Position instanceLine (length ("instance " :: String))
         range = Range startOfConstraint startOfConstraint
-        title = "Add `" <> constraint <> "` to the context of the instance declaration"
-        newConstraint = constraint <> " => "
-     in [(title, [TextEdit range newConstraint])]
+        newConstraint = missingConstraint <> " => "
+     in [(actionTitle missingConstraint, [TextEdit range newConstraint])]
 
 -- Suggests a constraint for an instance declaration with one or more existing constraints.
 -- • Could not deduce (Eq b) arising from a use of ‘==’
@@ -451,37 +458,40 @@ suggestInstanceConstraint contents Diagnostic {..}
 --   In the expression: x == y && x' == y'
 --   In an equation for ‘==’:
 --       (Pair x x') == (Pair y y') = x == y && x' == y'
-  | Just constraint <- findMissingConstraint _message
-  , Just [instanceLineStr, constraintFirstCharStr]
+  | Just [instanceLineStr, constraintFirstCharStr]
     <- matchRegex _message "bound by the instance declaration at .+:([0-9]+):([0-9]+)"
-  = let existingConstraints = secondLine _message
-        newConstraints = normalizeConstraints existingConstraints constraint
+  = let existingConstraints = findExistingConstraints _message
+        newConstraints = normalizeConstraints existingConstraints missingConstraint
         instanceLine = readPositionNumber instanceLineStr
         constraintFirstChar = readPositionNumber constraintFirstCharStr
         startOfConstraint = Position instanceLine constraintFirstChar
         endOfConstraint = Position instanceLine $
           constraintFirstChar + T.length existingConstraints
         range = Range startOfConstraint endOfConstraint
-        title = "Add `" <> constraint <> "` to the context of the instance declaration"
-     in [(title, [TextEdit range newConstraints])]
+     in [(actionTitle missingConstraint, [TextEdit range newConstraints])]
   | otherwise = []
     where
-      secondLine :: T.Text -> T.Text
-      secondLine = T.lines >>> flip (!!) 1 >>> T.strip >>> T.replace "from the context: " ""
+      findExistingConstraints :: T.Text -> T.Text
+      findExistingConstraints =
+        T.lines >>> flip (!!) 1 >>> T.strip >>> T.replace "from the context: " ""
 
       readPositionNumber :: T.Text -> Int
       readPositionNumber = T.unpack >>> read >>> pred
 
+      actionTitle :: T.Text -> T.Text
+      actionTitle constraint = "Add `" <> constraint
+        <> "` to the context of the instance declaration"
+
 findTypeSignatureName :: T.Text -> Maybe T.Text
-findTypeSignatureName t = matchRegex t "\\n\\s*(.+) :: " <&> head
+findTypeSignatureName t = matchRegex t "([^ ]+) :: " <&> head
 
 findTypeSignatureLine :: T.Text -> T.Text -> Int
 findTypeSignatureLine contents typeSignatureName =
   T.splitOn (typeSignatureName <> " :: ") contents & head & T.lines & length
 
 -- | Suggests a constraint for a type signature for which a constraint is missing.
-suggestFunctionConstraint :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
-suggestFunctionConstraint contents Diagnostic{..}
+suggestFunctionConstraint :: T.Text -> Diagnostic -> T.Text -> [(T.Text, [TextEdit])]
+suggestFunctionConstraint contents Diagnostic{..} missingConstraint
 -- Suggests a constraint for a type signature with any number of existing constraints.
 -- • No instance for (Eq a) arising from a use of ‘==’
 --   Possible fix:
@@ -504,19 +514,16 @@ suggestFunctionConstraint contents Diagnostic{..}
 --   In the expression: x == x' && y == y'
 --   In an equation for ‘eq’:
 --       eq (Pair x y) (Pair x' y') = x == x' && y == y'
-  | Just c <- contents
-  , True <- _message =~ ("the type signature for:" :: String)
-  , Just constraint <- findMissingConstraint _message
-  , Just typeSignatureName <- findTypeSignatureName _message
+  | Just typeSignatureName <- findTypeSignatureName _message
   = let mExistingConstraints = findExistingConstraints _message
-        newConstraint = buildNewConstraints constraint mExistingConstraints
-        typeSignatureLine = findTypeSignatureLine c typeSignatureName
+        newConstraint = buildNewConstraints missingConstraint mExistingConstraints
+        typeSignatureLine = findTypeSignatureLine contents typeSignatureName
         typeSignatureFirstChar = T.length $ typeSignatureName <> " :: "
         startOfConstraint = Position typeSignatureLine typeSignatureFirstChar
         endOfConstraint = Position typeSignatureLine $
           typeSignatureFirstChar + maybe 0 T.length mExistingConstraints
         range = Range startOfConstraint endOfConstraint
-     in [(actionTitle constraint typeSignatureName, [TextEdit range newConstraint])]
+     in [(actionTitle missingConstraint typeSignatureName, [TextEdit range newConstraint])]
   | otherwise = []
     where
       findExistingConstraints :: T.Text -> Maybe T.Text

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -1405,11 +1405,35 @@ addFunctionConstraintTests = let
     , "eq x y = x == y"
     ]
 
+  incompleteConstraintSourceCode :: Maybe T.Text -> T.Text
+  incompleteConstraintSourceCode mConstraint =
+    let constraint = maybe "Eq a" (\c -> "(Eq a, " <> c <> ")") mConstraint
+     in T.unlines
+    [ "module Testing where"
+    , ""
+    , "data Pair a b = Pair a b"
+    , ""
+    , "eq :: " <> constraint <> " => Pair a b -> Pair a b -> Bool"
+    , "eq (Pair x y) (Pair x' y') = x == x' && y == y'"
+    ]
+
+  incompleteConstraintSourceCode2 :: Maybe T.Text -> T.Text
+  incompleteConstraintSourceCode2 mConstraint =
+    let constraint = maybe "(Eq a, Eq b)" (\c -> "(Eq a, Eq b, " <> c <> ")") mConstraint
+     in T.unlines
+    [ "module Testing where"
+    , ""
+    , "data Three a b c = Three a b c"
+    , ""
+    , "eq :: " <> constraint <> " => Three a b c -> Three a b c -> Bool"
+    , "eq (Three x y z) (Three x' y' z') = x == x' && y == y' && z == z'"
+    ]
+
   check :: T.Text -> T.Text -> T.Text -> TestTree
   check actionTitle originalCode expectedCode = testSession (T.unpack actionTitle) $ do
     doc <- createDoc "Testing.hs" "haskell" originalCode
     _ <- waitForDiagnostics
-    actionsOrCommands <- getCodeActions doc (Range (Position 6 0) (Position 6 68))
+    actionsOrCommands <- getCodeActions doc (Range (Position 6 0) (Position 6 maxBound))
     chosenAction <- liftIO $ pickActionWithTitle actionTitle actionsOrCommands
     executeCodeAction chosenAction
     modifiedCode <- documentContents doc
@@ -1420,6 +1444,14 @@ addFunctionConstraintTests = let
     "Add `Eq a` to the context of the type signature for `eq`"
     (missingConstraintSourceCode Nothing)
     (missingConstraintSourceCode $ Just "Eq a")
+  , check
+    "Add `Eq b` to the context of the type signature for `eq`"
+    (incompleteConstraintSourceCode Nothing)
+    (incompleteConstraintSourceCode $ Just "Eq b")
+  , check
+    "Add `Eq c` to the context of the type signature for `eq`"
+    (incompleteConstraintSourceCode2 Nothing)
+    (incompleteConstraintSourceCode2 $ Just "Eq c")
   ]
 
 addSigActionTests :: TestTree

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -482,6 +482,7 @@ codeActionTests = testGroup "code actions"
   , insertNewDefinitionTests
   , deleteUnusedDefinitionTests
   , addInstanceConstraintTests
+  , addFunctionConstraintTests
   ]
 
 codeLensesTests :: TestTree
@@ -1377,7 +1378,7 @@ addInstanceConstraintTests = let
     modifiedCode <- documentContents doc
     liftIO $ expectedCode @=? modifiedCode
 
-  in testGroup "add constraint"
+  in testGroup "add instance constraint"
   [ check
     "Add `Eq a` to the context of the instance declaration"
     (missingConstraintSourceCode Nothing)
@@ -1390,6 +1391,35 @@ addInstanceConstraintTests = let
     "Add `Eq c` to the context of the instance declaration"
     (incompleteConstraintSourceCode2 Nothing)
     (incompleteConstraintSourceCode2 $ Just "Eq c")
+  ]
+
+addFunctionConstraintTests :: TestTree
+addFunctionConstraintTests = let
+  missingConstraintSourceCode :: Maybe T.Text -> T.Text
+  missingConstraintSourceCode mConstraint =
+    let constraint = maybe "" (<> " => ") mConstraint
+     in T.unlines
+    [ "module Testing where"
+    , ""
+    , "eq :: " <> constraint <> "a -> a -> Bool"
+    , "eq x y = x == y"
+    ]
+
+  check :: T.Text -> T.Text -> T.Text -> TestTree
+  check actionTitle originalCode expectedCode = testSession (T.unpack actionTitle) $ do
+    doc <- createDoc "Testing.hs" "haskell" originalCode
+    _ <- waitForDiagnostics
+    actionsOrCommands <- getCodeActions doc (Range (Position 6 0) (Position 6 68))
+    chosenAction <- liftIO $ pickActionWithTitle actionTitle actionsOrCommands
+    executeCodeAction chosenAction
+    modifiedCode <- documentContents doc
+    liftIO $ expectedCode @=? modifiedCode
+
+  in testGroup "add function constraint"
+  [ check
+    "Add `Eq a` to the context of the type signature for `eq`"
+    (missingConstraintSourceCode Nothing)
+    (missingConstraintSourceCode $ Just "Eq a")
   ]
 
 addSigActionTests :: TestTree

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -481,6 +481,7 @@ codeActionTests = testGroup "code actions"
   , addSigActionTests
   , insertNewDefinitionTests
   , deleteUnusedDefinitionTests
+  , addConstraintTests
   ]
 
 codeLensesTests :: TestTree
@@ -1327,6 +1328,35 @@ fillTypedHoleTests = let
           "_a" "_b"  "parameterInt"
 #endif
   ]
+
+addConstraintTests :: TestTree
+addConstraintTests = let
+  sourceCode :: Maybe T.Text -> T.Text
+  sourceCode mConstraint =
+    let constraint = maybe "" (\c -> "(" <> c <> ") => ") mConstraint
+     in T.unlines
+    [ "module Testing where"
+    , ""
+    , "data Wrap a = Wrap a"
+    , ""
+    , "instance " <> constraint <> "Eq (Wrap a) where"
+    , "  (Wrap x) == (Wrap y) = x == y"
+    ]
+
+  check :: T.Text -> TestTree
+  check actionTitle = testSession (T.unpack actionTitle) $ do
+    let originalCode = sourceCode Nothing
+    let expectedCode = sourceCode $ Just "Eq a"
+    doc <- createDoc "Testing.hs" "haskell" originalCode
+    _ <- waitForDiagnostics
+    actionsOrCommands <- getCodeActions doc (Range (Position 6 0) (Position 6 maxBound))
+    chosenAction <- liftIO $ pickActionWithTitle actionTitle actionsOrCommands
+    executeCodeAction chosenAction
+    modifiedCode <- documentContents doc
+    liftIO $ expectedCode @=? modifiedCode
+
+  in testGroup "add constraint"
+  [ check "Add `(Eq a)` to the context of the instance declaration" ]
 
 addSigActionTests :: TestTree
 addSigActionTests = let

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -481,7 +481,7 @@ codeActionTests = testGroup "code actions"
   , addSigActionTests
   , insertNewDefinitionTests
   , deleteUnusedDefinitionTests
-  , addConstraintTests
+  , addInstanceConstraintTests
   ]
 
 codeLensesTests :: TestTree
@@ -1329,11 +1329,11 @@ fillTypedHoleTests = let
 #endif
   ]
 
-addConstraintTests :: TestTree
-addConstraintTests = let
-  sourceCode :: Maybe T.Text -> T.Text
-  sourceCode mConstraint =
-    let constraint = maybe "" (\c -> "(" <> c <> ") => ") mConstraint
+addInstanceConstraintTests :: TestTree
+addInstanceConstraintTests = let
+  missingConstraintSourceCode :: Maybe T.Text -> T.Text
+  missingConstraintSourceCode mConstraint =
+    let constraint = maybe "" (<> " => ") mConstraint
      in T.unlines
     [ "module Testing where"
     , ""
@@ -1343,20 +1343,54 @@ addConstraintTests = let
     , "  (Wrap x) == (Wrap y) = x == y"
     ]
 
-  check :: T.Text -> TestTree
-  check actionTitle = testSession (T.unpack actionTitle) $ do
-    let originalCode = sourceCode Nothing
-    let expectedCode = sourceCode $ Just "Eq a"
+  incompleteConstraintSourceCode :: Maybe T.Text -> T.Text
+  incompleteConstraintSourceCode mConstraint =
+    let constraint = maybe "Eq a" (\c -> "(Eq a, " <> c <> ")")  mConstraint
+     in T.unlines
+    [ "module Testing where"
+    , ""
+    , "data Pair a b = Pair a b"
+    , ""
+    , "instance " <> constraint <> " => Eq (Pair a b) where"
+    , "  (Pair x y) == (Pair x' y') = x == x' && y == y'"
+    ]
+
+  incompleteConstraintSourceCode2 :: Maybe T.Text -> T.Text
+  incompleteConstraintSourceCode2 mConstraint =
+    let constraint = maybe "(Eq a, Eq b)" (\c -> "(Eq a, Eq b, " <> c <> ")")  mConstraint
+     in T.unlines
+    [ "module Testing where"
+    , ""
+    , "data Three a b c = Three a b c"
+    , ""
+    , "instance " <> constraint <> " => Eq (Three a b c) where"
+    , "  (Three x y z) == (Three x' y' z') = x == x' && y == y' && z == z'"
+    ]
+
+  check :: T.Text -> T.Text -> T.Text -> TestTree
+  check actionTitle originalCode expectedCode = testSession (T.unpack actionTitle) $ do
     doc <- createDoc "Testing.hs" "haskell" originalCode
     _ <- waitForDiagnostics
-    actionsOrCommands <- getCodeActions doc (Range (Position 6 0) (Position 6 maxBound))
+    actionsOrCommands <- getCodeActions doc (Range (Position 6 0) (Position 6 68))
     chosenAction <- liftIO $ pickActionWithTitle actionTitle actionsOrCommands
     executeCodeAction chosenAction
     modifiedCode <- documentContents doc
     liftIO $ expectedCode @=? modifiedCode
 
   in testGroup "add constraint"
-  [ check "Add `(Eq a)` to the context of the instance declaration" ]
+  [ check
+    "Add `Eq a` to the context of the instance declaration"
+    (missingConstraintSourceCode Nothing)
+    (missingConstraintSourceCode $ Just "Eq a")
+  , check
+    "Add `Eq b` to the context of the instance declaration"
+    (incompleteConstraintSourceCode Nothing)
+    (incompleteConstraintSourceCode $ Just "Eq b")
+  , check
+    "Add `Eq c` to the context of the instance declaration"
+    (incompleteConstraintSourceCode2 Nothing)
+    (incompleteConstraintSourceCode2 $ Just "Eq c")
+  ]
 
 addSigActionTests :: TestTree
 addSigActionTests = let

